hacep/src/main/java/org/kie/hacep/consumer/DroolsConsumerHandler.java

```diff
@@ -44,45 +45,85 @@
+        this.envConfig = envConfig;
+        this.sessionSnapShooter = new DefaultSessionSnapShooter(this.envConfig);
+        initializeKieSessionContext(this.envConfig);
         this.producer = producer;
-        commandHandler = new CommandHandler(kieSessionContext, config, producer, snapshooter);
-        if (config.isUnderTest()) {
+        this.commandHandler = new CommandHandler(this.kieSessionContext, this.envConfig, producer, this.sessionSnapShooter);
+        if (this.envConfig.isUnderTest()) {
             loggerForTest = PrinterUtil.getKafkaLoggerForTest(envConfig);
         }
     }
 
-    private void initializeKieSessionFromSnapshot(EnvConfig config) {
-        if(config.isSkipOnDemanSnapshot()) {// if true we reads the snapshots and wait until the first leaderElectionUpdate
-            this.infos = snapshooter.deserialize();
-            this.kieSessionContext = createSessionHolder(infos);
+    private void initializeKieSessionContext(EnvConfig config) {
+        if(config.isSkipOnDemandSnapshot()) {// if true we reads the snapshots and wait until the first leaderElectionUpdate
+            initializeSessionContextWithSnapshotCheck();
         } else{
-            kieSessionContext = new KieSessionContext();
-            createClasspathSession( kieSessionContext );
+            createAndInitializeSessionContextWithoutSnapshot();
+        }
+    }
+
+    private void initializeSessionContextWithSnapshotCheck() {
+        this.snapshotInfos = this.sessionSnapShooter.deserialize();
+        /* if the Snapshot is ok we use the KieContainer and KieSession from the infos,
+           otherwise (empty system) we create kiecontainer and kieSession from the envConfig's GAV */
+        if (this.snapshotInfos != null) {
+            initializeSessionContextFromSnapshot();
+        }else{
+            createAndInitializeSessionContextWithoutSnapshot();
+        }
+    }
+
+    private void createAndInitializeSessionContextWithoutSnapshot() {
+        KieServices srv = KieServices.get();
+        if (srv != null) {
+            KieContainer kieContainer;
+            if(envConfig.isUpdatableKJar()){
+                if (logger.isInfoEnabled()) {logger.info("Creating new KieContainer with KJar:{}", envConfig.getKjarGAV());}
+                String parts[] = envConfig.getKjarGAV().split(":");
+                kieContainer = srv.newKieContainer(srv.newReleaseId(parts[0], parts[1], parts[2]));
+                KieScanner scanner = srv.newKieScanner(kieContainer);
+                scanner.scanNow();
+            }else {
+                if (logger.isInfoEnabled()) { logger.info("Creating new Kie Session with the Kjar deployed");}
+                kieContainer = KieServices.get().newKieClasspathContainer();
+            }
+            this.kieSessionContext = new KieSessionContext();
+            this.kieSessionContext.init(kieContainer, kieContainer.newKieSession());
+        } else {
+            logger.error("KieService is null");
```

Comment:
This should be a (fatal) error. Please throw an exception and/or mark this node to be killed if it is not recoverable

Reply:
Ok